### ISSUE TITLE
Multicursor highlighting on selection behaves oddly

### ISSUE THREAD

**Issue by atombender** [Microsoft/vscode#21457]
- VSCode Version: 1.9.1

Steps to Reproduce:

1. Type the text `XXXXXX`
2. Select the first `X`
3. Hit cmd-D multiple times. The selection will grow exponentially:

![t](https://cloud.githubusercontent.com/assets/50314/23344401/af98e044-fc49-11e6-80f3-af1054ca8dc1.gif)

I expected each cmd-D keypress to select the next X.

Furthermore, the "highlight all" behaves similarly oddly: It will highlight _pairs_ of whatever you selected:

![t2](https://cloud.githubusercontent.com/assets/50314/23344473/096e8032-fc4b-11e6-93ee-3c7e1f90d68f.gif)

Even weirder, if you replace the text, then undo, any adjacent selections are collapsed (so it doesn't undo the selection properly):

![t3](https://cloud.githubusercontent.com/assets/50314/23351704/b173a006-fc90-11e6-9059-0d10959523d3.gif)

1. Type 20 Zs.
2. Select one Z and "highlight all". There are now 10 selections of 2 characters each (wrong).
3. Type "p" to replacing all of them. You get 10 p's (wrong).
4. Undo. There are now 5 selections of 4 characters each (wrong).
5. Type "p" to replace all of them again. You get 5 p's (wrong).
6. And so on.

**Comment by roblourens** [comment]
The first might be expected since contiguous selections are combined. Second looks like a bug?

**Comment by atombender** [comment]
Well, I did not expect the first one.

There are practical implications beyond just being weird: With the above behaviour it's impossible to select all Xs unless they occur in powers of 2. I.e. I can select 1, 2, 4, 8, ... Xs, but not, say, 10.

As for selections being combined: If I have `X X X X ...` (i.e. separated by spaces), it works: All instances of X are individually selected. Logically, it shouldn't matter if the Xs are next to each other or not, since my starting selection is a single character.

In other words, the behaviour is inconsistent. Not just that, it's arguably useless, because it's extremely rare to want to replace exactly 2<sup>n</sup> instances of anything.

VSC does seem capable of having multiple selections adjacent to each other, as shown by the second example (the Ys).

**Comment by atombender** [comment]
Updated issue with a GIF of the third problem (affecting "highlight all").

**Comment by martin-wiseweb** [comment]
I opened a similar issue... Possible duplicate of #11150 @alexandrudima

**Comment by atombender** [comment]
Looks like it's the same bug. Although mine demonstrates a few more edge cases that we'll want to keep in mind in whichever issue is kept open.

**Comment by paulpflug** [comment]
stumbled upon this, too. Usecase:
```
# replace all \n with newline
\n
\n\n
\n
```
currently not possible with ctrl+d, as selection changes from `\n` to `\n\n` on the second line

**Comment by alexandrudima** [comment]
Duplicate of #6661
